### PR TITLE
Tried to make it so that aif and em are interoperable. 

### PR DESCRIPTION
**Pull request by olliethomas**
Turns out that it's totally possible to switch between EM and AIF datasets, providing that you do a lot of the definitions at runtime.
The problem is that there are some fundamental design differences, and I don't know how to get past them, so putting this here so that everyone can see what I've done.

### CODE REVIEW

**Comment by olliethomas** [comment]
Is there a way to turn checks off? I know that this will fail.

@thomkeh 
couple things:
- the shape of the returned dataset is len(x) + len(s).
- I can drop the S feature and the shape will be right, but the bigger problem is that this shouldn't really be available (IMO) during training of the algorithm in the first place.

**Comment by thomkeh** [comment]
Okay, so the problem is that AIF360 expects s as part of the input features, but if you add s to x, then it is hard after the transformation to separate s and x again?

**Comment by olliethomas** [comment]
> Okay, so the problem is that AIF360 expects s as part of the input features, 
It does

> but if you add s to x, then it is hard after the transformation to separate s and x again?
No, it's possible. It's a bit awkward because you have a numpy array and a feature list, but that's not so much of a problem (assuming that the order of the list and the features align)

The problem is that the learning procedures in aif learns a mapping `f: (x,s) -> y`
In contrast we learn `f: x -> y`. While sometimes `s` is used during training e.g. in a loss function, we don't treat `s` as just another column in `x`
If the purpose of this is to be able to call AIF360 models from EM, then suddenly there is not a fair comparison.

**Comment by thomkeh** [comment]
Hm, this is an interesting question. We do make `s` available to the algorithms, I think (`TestTuple` has both `x` and `s`). But they usually don't make use of it. Why did we do it like that? :thinking:

**Comment by olliethomas** [comment]
Why do we do it like that as in, why do we _not_ use `s` as a feature?

Because outcomes shouldn't be based on sensitive attributes as to avoid claims of [disparate treatment](https://en.wikipedia.org/wiki/Disparate_treatment). Or this was the thinking anyway.

Any model that is trained as `(x,s) -> y` will also need `(x,s)` during inference.

File: ethicml/algorithms/preprocess/pre_algorithm.py [hunk 1]
```diff
@@ -32,27 +32,25 @@ def __init__(self, name: str, seed: int, out_size: Optional[int]):
         self._out_size = out_size
 
     @abstractmethod
-    def fit(self, train: DataTuple) -> Tuple[PreAlgorithm, DataTuple]:
```
Comment (thomkeh) [inline ethicml/algorithms/preprocess/pre_algorithm.py:35]: why was the return type like that?

**Review by thomkeh — COMMENTED** [review]
in general I have no objections to merging this. but if you think a better design is possible, we can hold off.

File: ethicml/algorithms/preprocess/pre_algorithm.py (hunk 1 above)
Comment (olliethomas) [reply ethicml/algorithms/preprocess/pre_algorithm.py:35]: 🤔  it should only really be the `PreAlgorithm`. I think I did it with both because the training set was always transformed so it feels like a waste to not return it... but this does make the api ugly...

Comment (thomkeh) [reply ethicml/algorithms/preprocess/pre_algorithm.py:35]: is the solution to introduce a `fit_and_transform()`?
